# **15 Ways to improve the ML Model's performance?**

 ---
### Author :  Muhammad Ali Saagar

**Python Developer • Data Science • AI/ML Engineer**

#### 🔗 Connect With Me

[![GitHub](https://img.shields.io/badge/GitHub-Profile-181717?style=for-the-badge\&logo=github)](https://github.com/AliMuhammad78)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-Profile-0A66C2?style=for-the-badge\&logo=linkedin)](https://www.linkedin.com/in/muhammad-ali-91294a290)
[![Kaggle](https://img.shields.io/badge/Kaggle-Profile-20BEFF?style=for-the-badge\&logo=kaggle)](https://www.kaggle.com/ali98muhammad45)
[![YouTube](https://img.shields.io/badge/YouTube-Channel-FF0000?style=for-the-badge\&logo=youtube)](https://www.youtube.com/@TechyBits20121)
[![Email](https://img.shields.io/badge/Email-Contact-D14836?style=for-the-badge\&logo=gmail)](mailto:alisaagar86@gmail.com)

### Acknowledgement

> Most of the examples in this notebook are based on the excellent tutorials by **Dr. Muhammad Aammar Tufail (Codanics)**. I sincerely thank him for sharing his knowledge and practical examples.

 [![GitHub](https://img.shields.io/badge/GitHub-Profile-181717?style=for-the-badge\&logo=github)](https://github.com/AammarTufail)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-Profile-0A66C2?style=for-the-badge\&logo=linkedin)](https://www.linkedin.com/in/dr-muhammad-aammar-tufail-02471213b/)
[![Kaggle](https://img.shields.io/badge/Kaggle-Profile-20BEFF?style=for-the-badge\&logo=kaggle)](https://www.kaggle.com/muhammadaammartufail)
[![YouTube](https://img.shields.io/badge/YouTube-Channel-FF0000?style=for-the-badge\&logo=youtube)](https://www.youtube.com/@codanics)

---

In [1]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder , StandardScaler , LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error , mean_absolute_percentage_error
from sklearn.model_selection import train_test_split

import numpy as np 
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt 

data = sns.load_dataset("diamonds")
X = data[["carat", "cut", "color", "clarity", "depth", "table", "x", "y", "z"]]
y = data["price"]

numeric_features = ["carat" , "depth" , "table" , "x" , "y" ,"z"]
categorical_features = ["cut" , "color" , "clarity"]

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)
    ]
)

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=42)

# pipeline
pipeline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('model', LinearRegression())
    ]
)

# fit the model
pipeline.fit(X_train, y_train)

# metric to evaluate the model
y_pred = pipeline.predict(X_test)

print(f"Mean Squared Error: {mean_squared_error(y_test, y_pred)}")
print(f"R2 Score: {r2_score(y_test, y_pred)}")
print(f"Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}")
print(f"Mean Absolute Percentage Error: {mean_absolute_percentage_error(y_test, y_pred)}")
# root mean squared error
print(f"Root Mean Squared Error: {np.sqrt(mean_squared_error(y_test, y_pred))}")

Mean Squared Error: 1288705.4778516763
R2 Score: 0.9189331350419386
Mean Absolute Error: 737.1513665933284
Mean Absolute Percentage Error: 0.3952933516494366
Root Mean Squared Error: 1135.2116445190634


# **Improving the model's performance**

## **15 Ways to improve the ML Model's performance?**

1. **Feature Engineering**: Create new features that capture additional information from the data.
2. **Hyperparameter Tuning**: Optimize the model's hyperparameters to improve performance.
3. **Regularization**: Apply regularization techniques to prevent overfitting.
4. **Ensemble Methods**: Use ensemble methods such as Random Forest or Gradient Boosting to improve predictive performance.
5. **Cross-Validation**: Use cross-validation to assess the model's performance more accurately.
6. **Feature Selection**: Identify and select the most relevant features for the model.
7. **Data Preprocessing**: Clean and preprocess the data to improve model performance.
8. **Model Selection**: Experiment with different regression models to find the best fit for the data.
9. **Error Analysis**: Analyze the model's errors to identify patterns and areas for improvement.
10. **Domain Knowledge**: Incorporate domain knowledge to improve the model's predictive power.
11. **Data Augmentation**: Increase the size of the training data through data augmentation techniques.
12. **Model Stacking**: Combine multiple models to improve predictive performance.
13. **Model Interpretation**: Interpret the model's predictions to gain insights into the data and improve performance.
14. **Model Deployment**: Deploy the model in a production environment and monitor its performance over time.
15. **Feedback Loop**: Incorporate feedback from users and stakeholders to continuously improve the model.

## Using Decision Tree Regressor 

In [2]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder , StandardScaler , LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error , mean_absolute_percentage_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor


import numpy as np 
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt 

data = sns.load_dataset("diamonds")
X = data[["carat", "cut", "color", "clarity", "depth", "table", "x", "y", "z"]]
y = data["price"]

numeric_features = ["carat" , "depth" , "table" , "x" , "y" ,"z"]
categorical_features = ["cut" , "color" , "clarity"]

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)
    ]
)

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=42)

# pipeline
pipeline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('model', DecisionTreeRegressor())
    ]
)

# fit the model
pipeline.fit(X_train, y_train)

# metric to evaluate the model
y_pred = pipeline.predict(X_test)

print(f"Mean Squared Error: {mean_squared_error(y_test, y_pred)}")
print(f"R2 Score: {r2_score(y_test, y_pred)}")
print(f"Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}")
print(f"Mean Absolute Percentage Error: {mean_absolute_percentage_error(y_test, y_pred)}")
# root mean squared error
print(f"Root Mean Squared Error: {np.sqrt(mean_squared_error(y_test, y_pred))}")

Mean Squared Error: 538430.6377224694
R2 Score: 0.966129666903958
Mean Absolute Error: 356.6618928439006
Mean Absolute Percentage Error: 0.08567647286718962
Root Mean Squared Error: 733.7783300987222


## Using Random Forest Regressor 

In [3]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder , StandardScaler , LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error , mean_absolute_percentage_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

import numpy as np 
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt 

data = sns.load_dataset("diamonds")
X = data[["carat", "cut", "color", "clarity", "depth", "table", "x", "y", "z"]]
y = data["price"]

numeric_features = ["carat" , "depth" , "table" , "x" , "y" ,"z"]
categorical_features = ["cut" , "color" , "clarity"]

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)
    ]
)

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=42)

# pipeline
pipeline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('model', RandomForestRegressor())
    ]
)

# fit the model
pipeline.fit(X_train, y_train)

# metric to evaluate the model
y_pred = pipeline.predict(X_test)

print(f"Mean Squared Error: {mean_squared_error(y_test, y_pred)}")
print(f"R2 Score: {r2_score(y_test, y_pred)}")
print(f"Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}")
print(f"Mean Absolute Percentage Error: {mean_absolute_percentage_error(y_test, y_pred)}")
# root mean squared error
print(f"Root Mean Squared Error: {np.sqrt(mean_squared_error(y_test, y_pred))}")

Mean Squared Error: 304874.556659746
R2 Score: 0.9808216656647679
Mean Absolute Error: 271.10742032262476
Mean Absolute Percentage Error: 0.0649677378633316
Root Mean Squared Error: 552.1544681153508


# Using XGBoost 

In [7]:
!pip install xgboost --quiet


[notice] A new release of pip is available: 24.2 -> 26.0.1
[notice] To update, run: C:\Users\zen lab\AppData\Local\Programs\Python\Python313\python.exe -m pip install --upgrade pip


In [10]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder , StandardScaler , LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error , mean_absolute_percentage_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor

import numpy as np 
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt 

data = sns.load_dataset("diamonds")
X = data[["carat", "cut", "color", "clarity", "depth", "table", "x", "y", "z"]]
y = data["price"]

numeric_features = ["carat" , "depth" , "table" , "x" , "y" ,"z"]
categorical_features = ["cut" , "color" , "clarity"]

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)
    ]
)

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=42)

# pipeline
pipeline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('model', XGBRegressor())
    ]
)

# fit the model
pipeline.fit(X_train, y_train)

# metric to evaluate the model
y_pred = pipeline.predict(X_test)

print(f"Mean Squared Error: {mean_squared_error(y_test, y_pred)}")
print(f"R2 Score: {r2_score(y_test, y_pred)}")
print(f"Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}")
print(f"Mean Absolute Percentage Error: {mean_absolute_percentage_error(y_test, y_pred)}")
# root mean squared error
print(f"Root Mean Squared Error: {np.sqrt(mean_squared_error(y_test, y_pred))}")

Mean Squared Error: 318286.3125
R2 Score: 0.9799779653549194
Mean Absolute Error: 285.61346435546875
Mean Absolute Percentage Error: 0.07398492097854614
Root Mean Squared Error: 564.1686915276316


# **Comparing the results** 

## Using Linear Regression 
- Mean Squared Error: 1288705.4778516763
- R2 Score: 0.9189331350419386
- Mean Absolute Error: 737.1513665933284
- Mean Absolute Percentage Error: 0.3952933516494366
- Root Mean Squared Error: 1135.2116445190634


## Using Decision Tree 

- Mean Squared Error: 538430.6377224694
- R2 Score: 0.966129666903958
- Mean Absolute Error: 356.6618928439006
- Mean Absolute Percentage Error: 0.08567647286718962
- Root Mean Squared Error: 733.7783300987222

## Using Random Forests 

- Mean Squared Error: 304874.556659746
- R2 Score: 0.9808216656647679
- Mean Absolute Error: 271.10742032262476
- Mean Absolute Percentage Error: 0.0649677378633316
- R- oot Mean Squared Error: 552.1544681153508

## Using XG Boost

- Mean Squared Error: 318286.3125
- R2 Score: 0.9799779653549194
- Mean Absolute Error: 285.61346435546875
- Root Mean Squared Error: 564.1686915276316
- Mean Absolute Percentage Error: 0.07398492097854614
